### PR TITLE
7738-view-data-not-populated

### PR DESCRIPTION
**Pull request by danarad05**
switching between view containers could cause view not to be populated with data

Signed-off-by: Dan Arad <dan.arad@sap.com>

#### What it does
Fixes: #7738

#### How to test
1. Load Theia when on `Explorer` tab and when "NPM scripts" section is expanded
image
2. before NPM scripts are populated with data, switch to "Search"  tab and let it finish loading
3. Return to "Explorer" tab - "NPM scripts" should populate shortly.

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by danarad05** [comment]
FYI @amiramw

**Comment by offer8** [comment]
hi @vince-fugnitto can we merge this one before the next release? 
It solves us very high bug in multiple stakeholder scenarios.
We verified that the fix solved the issue.

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts [hunk 1]
```diff
@@ -470,12 +470,13 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
                 part.onDidChangeVisibility(() => widget.suppressUpdateViewVisibility = part.isHidden);
 
                 const tryFireOnDidExpandView = () => {
-                    if (!part.collapsed && part.isVisible) {
+                    if (widget.widgets.length === 0) {
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:473]: @danarad05 can you explain what is happening here or why the change is necessary? 
The following fires the `onDidExpandViewEmitter` despite the view being collapsed.

**Comment by vince-fugnitto** [comment]
> hi @vince-fugnitto can we merge this one before the next release?

@offer8 I'll do my best to review but the pull-request is currently not approved. I confirmed it fixes the specific issue but I want to confirm that it is the proper solution. At the moment I see that the event is fired at all times, even if views are collapsed.

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts (hunk 1 above)
Comment (danarad05) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:473]: @vince-fugnitto 

A) The problem is that there are 2 parts for populating data of a certain view:
1.  onload `prepareViewContainer` is invoked which initiates process
https://github.com/eclipse-theia/theia/blob/eb83e0e00d6c87ca1a98e8a807e7d265b3263a01/packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts#L452

2.  When `TreeViewExtImpl` is created it invokes  `$registerTreeDataProvider` which invokes `registerViewDataProvider`
which retrieves relevant data and runs `prepareView` which in turn adds `viewDataWidget` to widgets array (= loads successfully).

https://github.com/eclipse-theia/theia/blob/eb83e0e00d6c87ca1a98e8a807e7d265b3263a01/packages/plugin-ext/src/main/browser/view/tree-views-main.ts#L57-L58 

B) Now - 
- If you load Theia and stay in `explorer` tab - you enter into first `if` - all is well. 
- If on the other hand you start on `explorer` and more to other tab - then you will enter `else` section - which means that it is waiting for `onDidExpandView` in order to complete loading of view with data.

image

C) So when you start with `explorer` and switch to another tab, `onDidExpandView` manages to be fired already but `registerViewDataProvider` isn't yet listening for it and so data is not populated.

image

D) I wanted to find an indicator if data was populated or not - and saw that `prepareView` is essentially adding viewDataWidget to widget when successful - so I based distinction on that .

So till widget is added the dataWidget - switching between tabs will invoke `onDidChangeVisibility` which will eventually fire `onDidExpandView` event - an upon it's firing - listeners (onCollapsed, onDidChangeVisibility) will be disposed and so - no additional events will be fired.

Hope I explained it clearly.

Comment (vince-fugnitto) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:473]: @danarad05 thanks for explaining, but with the updated changes we are now firing `onDidExpandViewEmitter` even if the view-part (ex: `npm`) is not expanded, is that correct?

---

I added logs and I see that the event is fired for the two use-cases which I believe is incorrect:
- the view is collapsed initially on startup:

```
root INFO --- firing 'onDidExpandViewEmitter' for id='npm', state={ collapsed=true, visible=true } ---
```

- the view is completely hidden on startup:

```
root INFO --- firing 'onDidExpandViewEmitter' for id='npm', state={ collapsed=true, visible=false } ---
```

Logically for both use-cases I do not see why firing `onDidExpandViewEmitter` makes sense (one is collapsed, and the other is completely hidden).

Comment (danarad05) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:473]: @vince-fugnitto 

You are right, it can be improved to filter out those scenarios. Fixed.

**Review by vince-fugnitto — APPROVED** [review]
I confirmed that the pull-request fixes the issue, and that the original logic for expanded state looks correct :+1:

**Comment by offer8** [comment]
thanks @vince-fugnitto  can it be part of 1.12.0 https://github.com/eclipse-theia/theia/milestone/17?

**Comment by offer8** [comment]
thanks @marechal-p !
